File: system-test/java/src/main/java/com/mycompany/myshop/testkit/dsl/core/shared/UseCaseContext.java
package com.mycompany.myshop.testkit.dsl.core.shared;

import com.mycompany.myshop.testkit.dsl.port.ExternalSystemMode;
import java.util.HashMap;
import java.util.Map;
import java.util.UUID;

public class UseCaseContext {
    private final ExternalSystemMode externalSystemMode;
    private final Map<String, String> paramMap;
    private final Map<String, String> resultMap;

    public UseCaseContext(ExternalSystemMode externalSystemMode) {
        this.externalSystemMode = externalSystemMode;
        this.paramMap = new HashMap<>();
        this.resultMap = new HashMap<>();
    }

    public ExternalSystemMode getExternalSystemMode() {
        return externalSystemMode;
    }

    public String getParamValue(String alias) {
        if(isNullOrBlank(alias)) {
            return alias;
        }

        if (paramMap.containsKey(alias)) {
            return paramMap.get(alias);
        }

        var value = generateParamValue(alias);
        paramMap.put(alias, value);

        return value;
    }

    public String getParamValueOrLiteral(String alias) {
        if(isNullOrBlank(alias)) {
            return alias;
        }

        return switch (externalSystemMode) {
            case STUB -> getParamValue(alias);
            case REAL -> alias;
            default -> throw new IllegalStateException("Unsupported external system mode: " + externalSystemMode);
        };
    }

    public void setResultEntry(String alias, String value) {
        ensureAliasNotNullBlank(alias);

        if (resultMap.containsKey(alias)) {
            throw new IllegalStateException("Alias already exists: " + alias);
        }

        resultMap.put(alias, value);
    }

    public void setResultEntryFailed(String alias, String errorMessage) {
        ensureAliasNotNullBlank(alias);
        setResultEntry(alias, "FAILED: " + errorMessage);
    }

    public String getResultValue(String alias) {
        if(isNullOrBlank(alias)) {
            return alias;
        }

        var value = resultMap.get(alias);
        if (value == null) {
            return alias; // Return literal value if not found as alias
        }

        if(value.contains("FAILED")) {
            throw new IllegalStateException("Cannot get result value for alias '" + alias + "' because the operation failed: " + value);
        }

        return value;
    }

    public String expandAliases(String message) {
        var expandedMessage = expandAlias(message, paramMap);
        expandedMessage = expandAlias(expandedMessage, resultMap);
        return expandedMessage;
    }

    private static String expandAlias(String message, Map<String, String> map) {
        var expandedMessage = message;
        for (var entry : map.entrySet()) {
            var alias = entry.getKey();
            var actualValue = entry.getValue();
            expandedMessage = expandedMessage.replace(alias, actualValue);
        }
        return expandedMessage;
    }

    private static String generateParamValue(String alias) {
        ensureAliasNotNullBlank(alias);
        var suffix = UUID.randomUUID().toString().substring(0, 8);
        return alias + "-" + suffix;
    }

    private static void ensureAliasNotNullBlank(String alias) {
        if (isNullOrBlank(alias)) {
            throw new IllegalArgumentException("Alias cannot be null or blank");
        }
    }

    private static boolean isNullOrBlank(String alias) {
        return alias == null || alias.isBlank();
    }
}

File: system-test/dotnet/Dsl.Core/Shared/UseCaseContext.cs
using Dsl.Port;
using System.Collections.Generic;


namespace Dsl.Core.Shared;

public class UseCaseContext
{
    private readonly Dictionary<string, string> _paramMap;
    private readonly Dictionary<string, string> _resultMap;

    public ExternalSystemMode ExternalSystemMode { get; }

    public UseCaseContext(ExternalSystemMode externalSystemMode)
    {
        ExternalSystemMode = externalSystemMode;
        _paramMap = new();
        _resultMap = new();
    }

    public string? GetParamValue(string? alias)
    {
        if (string.IsNullOrWhiteSpace(alias))
        {
            return alias;
        }

        if (_paramMap.TryGetValue(alias, out var value))
        {
            return value;
        }

        var generatedValue = GenerateParamValue(alias);
        _paramMap[alias] = generatedValue;

        return generatedValue;
    }

    public string? GetParamValueOrLiteral(string? alias)
    {
        if (string.IsNullOrWhiteSpace(alias))
        {
            return alias;
        }

        return ExternalSystemMode switch
        {
            ExternalSystemMode.Stub => GetParamValue(alias),
            ExternalSystemMode.Real => alias,
            _ => throw new InvalidOperationException($"Unsupported external system mode: {ExternalSystemMode}")
        };
    }

    public void SetResultEntry(string alias, string value)
    {
        EnsureAliasNotNullOrWhiteSpace(alias);

        if (_resultMap.ContainsKey(alias))
        {
            throw new InvalidOperationException($"Alias already exists: {alias}");
        }

        _resultMap[alias] = value;
    }

    public void SetResultEntryFailed(string alias, string errorMessage)
    {
        EnsureAliasNotNullOrWhiteSpace(alias);
        SetResultEntry(alias, $"FAILED: {errorMessage}");
    }

    public string? GetResultValue(string? alias)
    {
        if (string.IsNullOrWhiteSpace(alias))
        {
            return alias;
        }

        if (_resultMap.TryGetValue(alias, out var value))
        {
            if (value.Contains("FAILED"))
            {
                throw new InvalidOperationException($"Cannot get result value for alias '{alias}' because the operation failed: {value}");
            }
            return value;
        }

        return alias; // Return literal value if not found as alias
    }

    public string ExpandAliases(string message)
    {
        var expandedMessage = ExpandAlias(message, _paramMap);
        expandedMessage = ExpandAlias(expandedMessage, _resultMap);
        return expandedMessage;
    }

    private static string ExpandAlias(string message, Dictionary<string, string> map)
    {
        var expandedMessage = message;
        foreach (var entry in map)
        {
            var alias = entry.Key;
            var actualValue = entry.Value;
            expandedMessage = expandedMessage.Replace(alias, actualValue);
        }
        return expandedMessage;
    }

    private static string GenerateParamValue(string alias)
    {
        EnsureAliasNotNullOrWhiteSpace(alias);
        var suffix = Guid.NewGuid().ToString()[..8];
        return $"{alias}-{suffix}";
    }

    private static void EnsureAliasNotNullOrWhiteSpace(string alias)
    {
        if (string.IsNullOrWhiteSpace(alias))
        {
            throw new ArgumentException("Alias cannot be null or blank");
        }
    }
}
